chronos/asyncproc.nim

```diff
@@ -0,0 +1,1323 @@
+        # Close child process STDERR.
+        doCheck(posixSpawnFileActionsAddClose(actions, cint(2)))
+        # Make a duplicate of `stdoutHandle` as child process STDERR.
+        doCheck(posixSpawnFileActionsAddDup2(actions, cint(pipes.stdoutHandle),
+                                             cint(2)))
+        # Close child process side of `stdoutHandle`.
+        doCheck(posixSpawnFileActionsAddClose(actions,
+                                              cint(pipes.stdoutHandle)))
+        # Close parent process side of `stdoutHandle`.
+        if not(pipes.stdoutHolder.isEmpty()):
+          let fd = cint(pipes.stdoutHolder.fd())
+          doCheck(posixSpawnFileActionsAddClose(actions, fd))
+    ok(SpawnAttr(attrs: attrs, actions: actions))
+
+  proc free(v: var SpawnAttr): Result[void, OSErrorCode] =
+    block:
+      let res = posixSpawnAttrDestroy(v.attrs)
+      if res != 0:
+        discard posixSpawnFileActionsDestroy(v.actions)
+        return err(OSErrorCode(res))
+    block:
+      let res = posixSpawnFileActionsDestroy(v.actions)
+      if res != 0:
+        return err(OSErrorCode(res))
+    ok()
+
+  proc envToCStringArray(t: StringTableRef): cstringArray =
+    let itemsCount = len(t)
+    var
+      res = cast[cstringArray](alloc0((itemsCount + 1) * sizeof(cstring)))
+      i = 0
+    for key, value in pairs(t):
+      var x = key & "=" & value
+      res[i] = cast[cstring](alloc0(len(x) + 1))
+      copyMem(res[i], addr(x[0]), len(x))
+      inc(i)
+    res
+
+  proc envToCStringArray(): cstringArray =
+    let itemsCount =
+      block:
+        var res = 0
+        for key, value in envPairs(): inc(res)
+        res
+    var
+      res = cast[cstringArray](alloc0((itemsCount + 1) * sizeof(cstring)))
+      i = 0
+    for key, value in envPairs():
+      var x = key & "=" & value
+      res[i] = cast[cstring](alloc0(len(x) + 1))
+      copyMem(res[i], addr(x[0]), len(x))
+      inc(i)
+    res
+
+  when defined(macosx) or defined(macos) or defined(ios):
+    proc getEnvironment(): ptr cstringArray {.
+      importc: "_NSGetEnviron", header: "<crt_externs.h>".}
+  else:
+    var globalEnv {.importc: "environ", header: "<unistd.h>".}: cstringArray
+
+  proc getProcessEnvironment*(): StringTableRef =
+    var res = newStringTable(modeCaseInsensitive)
+    let env =
+      when defined(macosx) or defined(macos) or defined(ios):
+        getEnvironment()[]
+      else:
+        globalEnv
+    var i = 0
+    while not(isNil(env[i])):
+      let line = $env[i]
+      if len(line) > 0:
+        let delim = line.find('=')
+        if delim > 0:
+          res[substr(line, 0, delim - 1)] = substr(line, delim + 1)
+      inc(i)
+    res
+
+  proc getFullCommand(command: string, arguments: cstringArray): string =
+    let length =
+      if isNil(arguments):
+        0
+      else:
+        var res = 0
+        while not(isNil(arguments[res])): inc(res)
+        res
+    var res = @[command]
```

Comment:
Using `newSeqOfCap` here is an easy way to save some potential re-allocations in the loop below.

Technically, it's also possible to avoid many string and sequence allocations by directly copying the `cstrings` into a pre-allocated result string.